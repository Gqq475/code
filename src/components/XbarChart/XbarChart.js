import React from 'react'
import classes from './XbarChart.scss'
import Chart from 're-echarts'
import Utils from 'Utils'
type Props = {
  xbarData: Object,
  lineData: Object
};

const option = {
  tooltip: {
    trigger: 'axis',
    formatter: function (params) {
      var res = '<div><p>时间：' + params[3].data + '</p></div>'
      for (var i = 0; i < params.length - 1; i++) {
        res += '<p>' + params[i].seriesName + ':' + params[i].data + '</p>'
      }
      return res
    }
  },
  title: {
    text: 'XBAR Chart',
    textStyle: {
      color: '#337ab7'
    }
  },
  grid: {
    x: 100,
    y: 50,
    x2: 100,
    y2: 100
  },
  xAxis: [
    {
      type: 'category',
      boundaryGap: false,
      show: false,
      position: 'top',
      axisLabel: {
       //  X轴刻度配置
        rotate: 45,
        interval: 0 // 0：表示全部显示不间隔；auto:表示自动根据刻度个数和宽度自动设置间隔个数
      },
      data: []
    },
    {
      type: 'category',
      boundaryGap: false,
      position: 'bottom',
      axisLabel: {
       //  X轴刻度配置
        rotate: 45,
        interval: 0 // 0：表示全部显示不间隔；auto:表示自动根据刻度个数和宽度自动设置间隔个数
      },
      data: []
    }
  ],
  yAxis: [
    {
      type: 'value',
      min: 5.5,
      max: 8.5,
      splitLine: {show: false},
      // splitNumber: 1,
      axisLabel: {
        formatter: '{value}'
      }
    }
  ],
  series: [
    {
      name: '数值',
      type: 'line',
      symbol: 'circle',
      symbolSize: 7,
      itemStyle: {
        normal: {
          // color: 'green'
          color: function (params) {
            let color
            if (params.name === 'ooc') {
              color = 'red'
            } else {
              color = 'green'
            }
            return color
          }
        }
      },
      data: [''],
      lineStyle: {
        normal: {
          color: 'blue'   // 线条颜色
        }
      },
      markLine: {
        tooltip: {
          show: false
        },
        data: [
          {yAxis: 0, itemStyle: {normal: {color: 'blue', label: {formatter: 'UCL'}}}},
          {yAxis: 0, itemStyle: {normal: {color: 'red', label: {formatter: 'USL'}}}},
          {yAxis: 0, itemStyle: {normal: {color: '#000', label: {formatter: 'CL'}}}},
          {yAxis: 0, itemStyle: {normal: {color: '#000', label: {formatter: 'SL'}}}},
          {yAxis: 0, itemStyle: {normal: {color: 'red', label: {formatter: 'LSL'}}}},
          {yAxis: 0, itemStyle: {normal: {color: 'blue', label: {formatter: 'LCL'}}}}
        ],
        symbol: 'none',
        itemStyle: {
          normal: {
            lineStyle: {
              type: 'solid'
            }
          }
        }
      }
    },
    {
      name: 'X最大值',
      type: 'scatter',
      data: [''],
      symbol: 'none',
      itemStyle: {
        normal: {
          color: '#f7f7f7'
        }
      }
    },
    {
      name: 'X最小值',
      type: 'scatter',
      data: [''],
      symbol: 'none',
      itemStyle: {
        normal: {
          color: '#f7f7f7'
        }
      }
    },
    {
      name: '时间',
      type: 'line',
      data: []
    }
  ]
}

export class XbarChart extends React.Component {
  props: Props;
  constructor (props) {
    super(props)
    this.state = {
    }
  }
  render () {
    if (this.props.xbarData === [] ||
      this.props.xbarData === undefined ||
      this.props.xbarData === null ||
      !this.props.xbarData.length) {
      return (
        <div className={classes['DefectRate-container']}>
          <div style={{color: '#337ab7', fontSize: '16px', fontWeight: 'bold'}}>暂无数据</div>
        </div>
      )
    } else {
      let Count = []
      let Time = []
      let TypeStyle = []
      let Xmax = []
      let Xmin = []
      let ucl = this.props.lineData.ucl
      let usl = this.props.lineData.usl
      let sl = this.props.lineData.sl
      let cl = this.props.lineData.cl
      let lsl = this.props.lineData.lsl
      let lcl = this.props.lineData.lcl
      if (this.props.xbarData) {
        this.props.xbarData.forEach((item, i) => {
          Time.push(Utils.formatDate(new Date((item.dateTime))))
          Count.push(item.xbar.toFixed(2))
          TypeStyle.push(item.type)
          Xmax.push(item.xmax.toFixed(2))
          Xmin.push(item.xmin.toFixed(2))
        })
        for (let i = 0; i < TypeStyle.length; i++) {
          if (TypeStyle[i] === undefined) {
            TypeStyle[i] = ''
          }
          TypeStyle[1] = 'ooc'
          TypeStyle[0] = 'ooc'
        }
        option.xAxis[0].data = TypeStyle
        option.xAxis[1].data = Time
        option.series[0].data = Count
        option.series[1].data = Xmax
        option.series[2].data = Xmin
        option.series[3].data = Time
        option.series[0].markLine.data[0].yAxis = ucl
        option.series[0].markLine.data[1].yAxis = usl
        option.series[0].markLine.data[2].yAxis = sl
        option.series[0].markLine.data[3].yAxis = cl
        option.series[0].markLine.data[4].yAxis = lsl
        option.series[0].markLine.data[5].yAxis = lcl
        return (
          <div className={classes['XBarChart-container']}>
            <Chart style={{position: 'relative', height: '450px'}} option={option} />
            <div style={{textAlign: 'center'}}>X轴： 时间</div>
            <div className={classes['yy']} style={{position: 'absolute', top: '25%'}}>Y轴： 数值</div>
          </div>
        )
      }
    }
  }
}
export default XbarChart
