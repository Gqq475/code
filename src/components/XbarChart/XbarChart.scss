.XbarChart-container {
	width: '100%'
}
.yy{
    transform:rotate(270deg);
    -ms-transform:rotate(270deg); /* <a href="https://www.baidu.com/s?wd=Internet&tn=44039180_cpr&fenlei=mv6quAkxTZn0IZRqIHckPjm4nH00T1Y3n1Rsnym4rjI-nHcsryc40ZwV5Hcvrjm3rH6sPfKWUMw85HfYnjn4nH6sgvPsT6KdThsqpZwYTjCEQLGCpyw9Uz4Bmy-bIi4WUvYETgN-TLwGUv3EPHc4rHDYP10s" target="_blank" class="baidu-highlight">Internet</a> <a href="https://www.baidu.com/s?wd=Explorer&tn=44039180_cpr&fenlei=mv6quAkxTZn0IZRqIHckPjm4nH00T1Y3n1Rsnym4rjI-nHcsryc40ZwV5Hcvrjm3rH6sPfKWUMw85HfYnjn4nH6sgvPsT6KdThsqpZwYTjCEQLGCpyw9Uz4Bmy-bIi4WUvYETgN-TLwGUv3EPHc4rHDYP10s" target="_blank" class="baidu-highlight">Explorer</a> 9*/
    -moz-transform:rotate(90deg); /* Firefox */
    -webkit-transform:rotate(270deg); /* Safari å’Œ Chrome */
    -o-transform:rotate(270deg); /* Opera */
    filter:progid:DXImageTransform.Microsoft.BasicImage(rotation=1); 
}