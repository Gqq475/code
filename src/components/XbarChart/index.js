import XbarChart from './XbarChart'
export default XbarChart
