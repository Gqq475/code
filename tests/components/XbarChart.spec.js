import React from 'react'
import XbarChart from 'components/XbarChart/XbarChart'

describe('(Component) XbarChart', () => {
  it('if Xbarchart "did" not exist ', () => {
    const forEach = jest.fn()
    const data = {
      lineData: {
        'ucl': 8.02,
        'lcl': 5.99,
        'cl': 7,
        'usl': 7.5,
        'sl': 7,
        'lsl': 6.5
      },
      xbarData: [
        {
          'dateTime': '2017/04/05 16:25:00',
          'xbar': 7.42,
          'xmax': 7.8,
          'xmin': 7.03,
          'type': 'ooc',
          'againstRules': [
            '規則1: 連續一點超出管制線外',
            '發生SOOS單點超規'
          ]
        },
        {
          'dateTime': '2017/04/05 16:25:00',
          'xbar': 7.42,
          'xmax': 7.8,
          'xmin': 7.03,
          'type': 'oos',
          'againstRules': [
            '規則1: 連續一點超出管制線外',
            '發生SOOS單點超規'
          ]
        }
      ]
    }
    const wrapper = shallow(
      <XbarChart {...data}
        forEach={forEach}
      />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  it('if Xbarchart "did" exist ', () => {
    const data = {}
    const wrapper = shallow(
      <XbarChart data={data} />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
