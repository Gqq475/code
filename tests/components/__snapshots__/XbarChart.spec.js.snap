// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) XbarChart if Xbarchart "did" exist  1`] = `
<div
  className="DefectRate-container"
>
  <div
    style={
      Object {
        "color": "#337ab7",
        "fontSize": "16px",
        "fontWeight": "bold",
      }
    }
  >
    暂无数据
  </div>
</div>
`;

exports[`(Component) XbarChart if Xbarchart "did" not exist  1`] = `
<div
  className="XBarChart-container"
>
  <ECharts
    config={Object {}}
    id="chart"
    lazyUpdate={false}
    notMerge={false}
    option={
      Object {
        "grid": Object {
          "x": 100,
          "x2": 100,
          "y": 50,
          "y2": 100,
        },
        "series": Array [
          Object {
            "data": Array [
              "7.42",
              "7.42",
            ],
            "itemStyle": Object {
              "normal": Object {
                "color": [Function],
              },
            },
            "lineStyle": Object {
              "normal": Object {
                "color": "blue",
              },
            },
            "markLine": Object {
              "data": Array [
                Object {
                  "itemStyle": Object {
                    "normal": Object {
                      "color": "blue",
                      "label": Object {
                        "formatter": "UCL",
                      },
                    },
                  },
                  "yAxis": 8.02,
                },
                Object {
                  "itemStyle": Object {
                    "normal": Object {
                      "color": "red",
                      "label": Object {
                        "formatter": "USL",
                      },
                    },
                  },
                  "yAxis": 7.5,
                },
                Object {
                  "itemStyle": Object {
                    "normal": Object {
                      "color": "#000",
                      "label": Object {
                        "formatter": "CL",
                      },
                    },
                  },
                  "yAxis": 7,
                },
                Object {
                  "itemStyle": Object {
                    "normal": Object {
                      "color": "#000",
                      "label": Object {
                        "formatter": "SL",
                      },
                    },
                  },
                  "yAxis": 7,
                },
                Object {
                  "itemStyle": Object {
                    "normal": Object {
                      "color": "red",
                      "label": Object {
                        "formatter": "LSL",
                      },
                    },
                  },
                  "yAxis": 6.5,
                },
                Object {
                  "itemStyle": Object {
                    "normal": Object {
                      "color": "blue",
                      "label": Object {
                        "formatter": "LCL",
                      },
                    },
                  },
                  "yAxis": 5.99,
                },
              ],
              "itemStyle": Object {
                "normal": Object {
                  "lineStyle": Object {
                    "type": "solid",
                  },
                },
              },
              "symbol": "none",
              "tooltip": Object {
                "show": false,
              },
            },
            "name": "数值",
            "symbol": "circle",
            "symbolSize": 7,
            "type": "line",
          },
          Object {
            "data": Array [
              "7.80",
              "7.80",
            ],
            "itemStyle": Object {
              "normal": Object {
                "color": "#f7f7f7",
              },
            },
            "name": "X最大值",
            "symbol": "none",
            "type": "scatter",
          },
          Object {
            "data": Array [
              "7.03",
              "7.03",
            ],
            "itemStyle": Object {
              "normal": Object {
                "color": "#f7f7f7",
              },
            },
            "name": "X最小值",
            "symbol": "none",
            "type": "scatter",
          },
          Object {
            "data": Array [
              "2017-04-05 08:25:00",
              "2017-04-05 08:25:00",
            ],
            "name": "时间",
            "type": "line",
          },
        ],
        "title": Object {
          "text": "XBAR Chart",
          "textStyle": Object {
            "color": "#337ab7",
          },
        },
        "tooltip": Object {
          "formatter": [Function],
          "trigger": "axis",
        },
        "xAxis": Array [
          Object {
            "axisLabel": Object {
              "interval": 0,
              "rotate": 45,
            },
            "boundaryGap": false,
            "data": Array [
              "ooc",
              "ooc",
            ],
            "position": "top",
            "show": false,
            "type": "category",
          },
          Object {
            "axisLabel": Object {
              "interval": 0,
              "rotate": 45,
            },
            "boundaryGap": false,
            "data": Array [
              "2017-04-05 08:25:00",
              "2017-04-05 08:25:00",
            ],
            "position": "bottom",
            "type": "category",
          },
        ],
        "yAxis": Array [
          Object {
            "axisLabel": Object {
              "formatter": "{value}",
            },
            "max": 8.5,
            "min": 5.5,
            "splitLine": Object {
              "show": false,
            },
            "type": "value",
          },
        ],
      }
    }
    style={
      Object {
        "height": "450px",
        "position": "relative",
      }
    }
  />
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    X轴： 时间
  </div>
  <div
    className="yy"
    style={
      Object {
        "position": "absolute",
        "top": "25%",
      }
    }
  >
    Y轴： 数值
  </div>
</div>
`;
